client/src/components/PresentedOptions.js

```diff
@@ -0,0 +1,504 @@
+    const Datastore = CKANApi.loadDatastore(endpoint, resource);
+
+    // We'll loop through the selected columns an check if we've loaded values for fields previously
+    selectedFields.forEach(field => {
+      // Check the cache
+      if (suggestedOptionCache[field]) {
+        options = options.concat(suggestedOptionCache[field]);
+        return;
+      }
+
+      // Start loading and append the promise to an array so we can wait for all of them...
+      loadPromises.push(
+        Datastore.search([field], null, true).then(result => {
+          let newOptions = [];
+
+          // TODO implement something for if the request fails...
+          if (result) {
+            newOptions = result.records.map(record => record[field]);
+          }
+
+          // Update the "cache" with the options we've loaded.
+          this.setState(state => ({
+            suggestedOptionCache: {
+              // We can't use the local variable as the state may have been mutated while this
+              // promise was resolving...
+              ...state.suggestedOptionCache,
+              [field]: newOptions,
+            },
+          }));
+
+          // Wait for any other suggested options to load from CKAN
+          // TODO this is _not_ a robust method of doing this. Preferably we can wait for all the
+          // promises to complete - or chain the promises. Note that the user has to be pretty quick
+          // to choose two new columns and beat the snappy response from CKAN. Perhaps we can
+          // disable the select while this updates?
+          setTimeout(() => { this.loadSuggestedOptions(); }, 1000);
+        })
+      );
+    });
+
+    // If we didn't have to load options then we can just put the known options into state
+    if (!loadPromises.length) {
+      options = this.splitOptionsBySeparators(options, this.state.usedSeparators);
+      const suggestedOptions = options
+        // Trim whitespace and convert all whitespace chunks to a single space
+        .map(item => item.trim().replace(/\s+/g, ' '))
+        .filter((item, index) => {
+          // Exclude null, non-string or empty values
+          if (!item || typeof item !== 'string' || item.length === 0) {
+            return false;
+          }
+          // Exclude items that aren't unique. AKA remove if this index is not the same as the
+          // index where this item is first found
+          if (options.indexOf(item) !== index) {
+            return false;
+          }
+
+          return true;
+        })
+        .sort();
+
+      this.setState({
+        // Trim, filter nulls, unique and sort the options...
+        suggestedOptions,
+        loading: false,
+      });
+      return;
+    }
+
+    // Mark as loading
+    this.setState({
+      loading: true,
+    });
+
+    // We have to wait for all promises and then just run this function again...
+    // Promise.all(loadPromises).then(() => this.loadSuggestedOptions());
```

Comment:
I reckon this function could be broken up a bit - thoughts?

Reply:
Yep it was getting a bit bloated. I split the `fetch` into it's own method and also the process that trims and filters the suggested options.